## Projects:

- Diffusion Models;
- YOLO versions  - object detection;
- Segment Anything Model (https://segment-anything.com/) (Liked);
- Mathematical Integration;
- BERT fine-tuning - Text classification (compare with logistic regression and TF-IDF vectors);
- CNN visualization - review the different techniques used to analyze CNN predictions;
- Monte Carlo dropout - review briefly the concept of uncertainty and what are the main techniques used to deal with uncertainty in deep learning;
- Deep vs Wide Neural Networks - Try to understand the effect of adding layers or adding parameters in a layer;
- Overparameterization regime;
- Neural ODEs - Neural Networks and ordinary differential equations are the two main modelization techniques used nowadays, review papers explaining some evolution of the idea of Neural ODE.

General observations:

- Presentation of 15 min 
- 70% grade
- No need to submit the code

Evaluation aspects:

- Clarity
- Understanding of concepts
- Depth of research and/or experiments

In [2]:
import torch
import numpy as np
import cv2
import os
import pandas as pd
import matplotlib.pyplot as plt
from segment_anything import SamPredictor, sam_model_registry
from sklearn.model_selection import train_test_split
from tqdm import tqdm


# Load SAM model
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
sam = sam_model_registry["vit_h"](checkpoint="C:/Users/elisa/OneDrive/Desktop/Deep Learning/DL_Project/sam_vit_h_4b8939.pth").to(device)
predictor = SamPredictor(sam)
print("✅ SAM Model loaded successfully!")

# Path to dataset
dataset_path = "C:/Users/elisa/OneDrive/Desktop/Deep Learning/DL_Project/Deep_Learning_FinalProject/Breast_Images/"


c:\Users\elisa\anaconda3\Lib\site-packages\segment_anything\build_sam.py:105: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  state_dict = torch.load(f)


✅ SAM Model loaded successfully!


In [9]:
# Path to dataset
dataset_path = "C:/Users/elisa/OneDrive/Desktop/Deep Learning/DL_Project/Deep_Learning_FinalProject/Breast_Images/"

# Define the class mapping
class_map = {"normal": 0, "benign": 1, "malignant": 2}

# List to store dataset entries
data = []

# Loop through each category
for category in ["normal", "benign", "malignant"]:
    category_path = os.path.join(dataset_path, category)
    
    # Get all image files in the category folder
    for file_name in os.listdir(category_path):
        if file_name.endswith(('.png', '.jpg', '.jpeg')):  # Check if it's an image
            image_path = os.path.join(category_path, file_name)
            data.append({"image_path": image_path, "label": class_map[category]})

# Convert to Pandas DataFrame
df = pd.DataFrame(data)

In [17]:
df

image_path  label
0     C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      0
1     C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      0
2     C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      0
3     C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      0
4     C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      0
...                                                 ...    ...
1352  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1353  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1354  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1355  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1356  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2

[1357 rows x 2 columns]

In [42]:
df.shape

(1357, 2)

We had to manually remove some images from the dataset as they contained human input identifying where the tumor is. In this way, we avoided training the CNN with image modified manually to indicate where the tumor is. 
From 1578 now we have 1357, so we lost 14% of our dataset.

The code below takes the ground thruth information created by the doctor and draw a bouding box around the actual breast ultrasound (adding some pixels to consdier a bigger area). It then crops the image around the bounding box creating ad hoc images of the turmor area. 
The code takes into account the presence of multiple masks for each image creating multiple cropped images for the same tumor image.
We created random bounding box to create random images of normal ultrasound to train our CNN to recognize when there is no tumor.

In [43]:
# Define output directories for cropped tumors
output_dir = "Cropped_Tumors"
benign_dir = os.path.join(output_dir, "benign")
malignant_dir = os.path.join(output_dir, "malignant")
normal_dir = os.path.join(output_dir, "normal")

os.makedirs(benign_dir, exist_ok=True)
os.makedirs(malignant_dir, exist_ok=True)
os.makedirs(normal_dir, exist_ok=True)

# Create a dictionary to store multiple masks for each image
image_mask_pairs = {}

# Store last bounding box from benign/malignant images
last_bbox = None  

# Iterate through the DataFrame to pair images and masks
for index, row in df.iterrows():
    image_path = row["image_path"]
    
    if "_mask" in image_path:  # This is a mask
        # Extract the base image path without any mask suffix
        base_image_path = (
            image_path.replace("_mask_2.png", ".png")
                      .replace("_mask_1.png", ".png")
                      .replace("_mask.png", ".png")
        )
        
        if base_image_path not in image_mask_pairs:
            image_mask_pairs[base_image_path] = []
        
        # Store all available masks in a list
        image_mask_pairs[base_image_path].append(image_path)

# Process benign and malignant images first
for index, row in tqdm(df.iterrows(), total=len(df)):
    image_path = row["image_path"]
    label = row["label"]  # 1 = benign, 2 = malignant

    # Skip normal images (they will be processed later)
    if label == 0 or "_mask" in image_path:
        continue  

    # Find corresponding masks (if any)
    mask_paths = sorted(image_mask_pairs.get(image_path, []))  

    # Load the ultrasound image
    image = cv2.imread(image_path)
    if image is None:
        print(f"Warning: Could not load {image_path}")
        continue
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image = cv2.resize(image, (1024, 1024))  # Resize for consistency

    # Process each mask for the image
    for mask_index, mask_path in enumerate(mask_paths):
        # Load the mask image
        mask = cv2.imread(mask_path, cv2.IMREAD_GRAYSCALE)
        if mask is None:
            print(f"Warning: Could not load {mask_path}")
            continue
        mask = cv2.resize(mask, (1024, 1024))  # Ensure same size as image

        # Find contours in the mask
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        if len(contours) == 0:
            print(f"No tumor detected in {mask_path}")
            continue

        # Select the largest contour (assuming it's the tumor)
        largest_contour = max(contours, key=cv2.contourArea)

        # Get bounding box around the largest contour
        x, y, w, h = cv2.boundingRect(largest_contour)
        last_bbox = (x, y, w, h)  # Store last bounding box

        # Extract the tumor region
        tumor_crop = image[y:y+h, x:x+w]

        # Select folder based on label
        label_dir = benign_dir if label == 1 else malignant_dir

        # Save the cropped tumor image with mask index
        tumor_filename = os.path.join(label_dir, f"tumor_{index}_mask{mask_index}.png")
        cv2.imwrite(tumor_filename, cv2.cvtColor(tumor_crop, cv2.COLOR_RGB2BGR))

        print(f"Saved: {tumor_filename}")

# Now process only normal ultrasound images using the last known bounding box
for index, row in tqdm(df.iterrows(), total=len(df)):
    image_path = row["image_path"]
    label = row["label"]  # 0 = normal

    # Process only normal ultrasound images (skip masks)
    if label != 0 or "_mask" in image_path:
        continue  

    # Load the ultrasound image
    image = cv2.imread(image_path)
    if image is None:
        print(f"Warning: Could not load {image_path}")
        continue
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image = cv2.resize(image, (1024, 1024))  # Resize for consistency

    # If no bounding box has been recorded, skip the normal image
    if last_bbox is None:
        print(f"Skipping {image_path}: No bounding box available from previous tumors.")
        continue

    # Crop using the last bounding box
    x, y, w, h = last_bbox
    normal_crop = image[y:y+h, x:x+w]

    # Save the cropped normal image
    normal_filename = os.path.join(normal_dir, f"normal_{index}.png")
    cv2.imwrite(normal_filename, cv2.cvtColor(normal_crop, cv2.COLOR_RGB2BGR))

    print(f"Saved: {normal_filename}")

print("Processing complete. Cropped tumor images and normal images are saved in:", output_dir)


 20%|█▉        | 271/1357 [00:00<00:00, 2650.01it/s]

Saved: Cropped_Tumors\benign\tumor_266_mask0.png
Saved: Cropped_Tumors\benign\tumor_268_mask0.png
Saved: Cropped_Tumors\benign\tumor_270_mask0.png
Saved: Cropped_Tumors\benign\tumor_270_mask1.png
Saved: Cropped_Tumors\benign\tumor_273_mask0.png
Saved: Cropped_Tumors\benign\tumor_275_mask0.png
Saved: Cropped_Tumors\benign\tumor_277_mask0.png
Saved: Cropped_Tumors\benign\tumor_279_mask0.png
Saved: Cropped_Tumors\benign\tumor_281_mask0.png
Saved: Cropped_Tumors\benign\tumor_283_mask0.png
Saved: Cropped_Tumors\benign\tumor_285_mask0.png
Saved: Cropped_Tumors\benign\tumor_287_mask0.png
Saved: Cropped_Tumors\benign\tumor_289_mask0.png
Saved: Cropped_Tumors\benign\tumor_291_mask0.png
Saved: Cropped_Tumors\benign\tumor_293_mask0.png
Saved: Cropped_Tumors\benign\tumor_295_mask0.png
Saved: Cropped_Tumors\benign\tumor_297_mask0.png
Saved: Cropped_Tumors\benign\tumor_299_mask0.png
Saved: Cropped_Tumors\benign\tumor_301_mask0.png
Saved: Cropped_Tumors\benign\tumor_303_mask0.png
Saved: Cropped_Tumor

 40%|███▉      | 537/1357 [00:03<00:06, 120.15it/s] 

Saved: Cropped_Tumors\benign\tumor_518_mask0.png
Saved: Cropped_Tumors\benign\tumor_520_mask0.png
Saved: Cropped_Tumors\benign\tumor_522_mask0.png
Saved: Cropped_Tumors\benign\tumor_524_mask0.png
Saved: Cropped_Tumors\benign\tumor_526_mask0.png
Saved: Cropped_Tumors\benign\tumor_528_mask0.png
Saved: Cropped_Tumors\benign\tumor_530_mask0.png
Saved: Cropped_Tumors\benign\tumor_532_mask0.png
Saved: Cropped_Tumors\benign\tumor_532_mask1.png
Saved: Cropped_Tumors\benign\tumor_535_mask0.png
Saved: Cropped_Tumors\benign\tumor_537_mask0.png
Saved: Cropped_Tumors\benign\tumor_539_mask0.png
Saved: Cropped_Tumors\benign\tumor_541_mask0.png
Saved: Cropped_Tumors\benign\tumor_543_mask0.png
Saved: Cropped_Tumors\benign\tumor_545_mask0.png
Saved: Cropped_Tumors\benign\tumor_547_mask0.png
Saved: Cropped_Tumors\benign\tumor_549_mask0.png
Saved: Cropped_Tumors\benign\tumor_551_mask0.png
Saved: Cropped_Tumors\benign\tumor_553_mask0.png
Saved: Cropped_Tumors\benign\tumor_555_mask0.png
Saved: Cropped_Tumor

 48%|████▊     | 651/1357 [00:05<00:06, 103.81it/s]

Saved: Cropped_Tumors\benign\tumor_642_mask0.png
Saved: Cropped_Tumors\benign\tumor_644_mask0.png
Saved: Cropped_Tumors\benign\tumor_646_mask0.png
Saved: Cropped_Tumors\benign\tumor_648_mask0.png
Saved: Cropped_Tumors\benign\tumor_650_mask0.png
Saved: Cropped_Tumors\benign\tumor_652_mask0.png
Saved: Cropped_Tumors\benign\tumor_654_mask0.png
Saved: Cropped_Tumors\benign\tumor_656_mask0.png
Saved: Cropped_Tumors\benign\tumor_658_mask0.png
Saved: Cropped_Tumors\benign\tumor_660_mask0.png
Saved: Cropped_Tumors\benign\tumor_662_mask0.png
Saved: Cropped_Tumors\benign\tumor_664_mask0.png
Saved: Cropped_Tumors\benign\tumor_666_mask0.png
Saved: Cropped_Tumors\benign\tumor_668_mask0.png
Saved: Cropped_Tumors\benign\tumor_670_mask0.png
Saved: Cropped_Tumors\benign\tumor_670_mask1.png
Saved: Cropped_Tumors\benign\tumor_673_mask0.png
Saved: Cropped_Tumors\benign\tumor_675_mask0.png
Saved: Cropped_Tumors\benign\tumor_677_mask0.png
Saved: Cropped_Tumors\benign\tumor_679_mask0.png
Saved: Cropped_Tumor

 53%|█████▎    | 717/1357 [00:06<00:06, 96.16it/s] 

Saved: Cropped_Tumors\benign\tumor_705_mask0.png
Saved: Cropped_Tumors\benign\tumor_707_mask0.png
Saved: Cropped_Tumors\benign\tumor_709_mask0.png
Saved: Cropped_Tumors\benign\tumor_711_mask0.png
Saved: Cropped_Tumors\benign\tumor_713_mask0.png
Saved: Cropped_Tumors\benign\tumor_715_mask0.png
Saved: Cropped_Tumors\benign\tumor_717_mask0.png
Saved: Cropped_Tumors\benign\tumor_719_mask0.png
Saved: Cropped_Tumors\benign\tumor_721_mask0.png
Saved: Cropped_Tumors\benign\tumor_723_mask0.png
Saved: Cropped_Tumors\benign\tumor_725_mask0.png
Saved: Cropped_Tumors\benign\tumor_727_mask0.png
Saved: Cropped_Tumors\benign\tumor_729_mask0.png
Saved: Cropped_Tumors\benign\tumor_731_mask0.png
Saved: Cropped_Tumors\benign\tumor_733_mask0.png
Saved: Cropped_Tumors\benign\tumor_735_mask0.png
Saved: Cropped_Tumors\benign\tumor_737_mask0.png
Saved: Cropped_Tumors\benign\tumor_739_mask0.png
Saved: Cropped_Tumors\benign\tumor_741_mask0.png
Saved: Cropped_Tumors\benign\tumor_743_mask0.png
Saved: Cropped_Tumor

 56%|█████▌    | 760/1357 [00:06<00:06, 88.61it/s]

Saved: Cropped_Tumors\benign\tumor_747_mask0.png
Saved: Cropped_Tumors\benign\tumor_749_mask0.png
Saved: Cropped_Tumors\benign\tumor_751_mask0.png
Saved: Cropped_Tumors\benign\tumor_753_mask0.png
Saved: Cropped_Tumors\benign\tumor_755_mask0.png
Saved: Cropped_Tumors\benign\tumor_757_mask0.png
Saved: Cropped_Tumors\benign\tumor_759_mask0.png
Saved: Cropped_Tumors\benign\tumor_761_mask0.png
Saved: Cropped_Tumors\benign\tumor_763_mask0.png
Saved: Cropped_Tumors\benign\tumor_765_mask0.png
Saved: Cropped_Tumors\benign\tumor_767_mask0.png
Saved: Cropped_Tumors\benign\tumor_769_mask0.png
Saved: Cropped_Tumors\benign\tumor_771_mask0.png
Saved: Cropped_Tumors\benign\tumor_773_mask0.png
Saved: Cropped_Tumors\benign\tumor_775_mask0.png
Saved: Cropped_Tumors\benign\tumor_777_mask0.png
Saved: Cropped_Tumors\benign\tumor_779_mask0.png
Saved: Cropped_Tumors\benign\tumor_781_mask0.png
Saved: Cropped_Tumors\benign\tumor_783_mask0.png
Saved: Cropped_Tumors\benign\tumor_785_mask0.png


 58%|█████▊    | 790/1357 [00:07<00:06, 82.41it/s]

Saved: Cropped_Tumors\benign\tumor_787_mask0.png
Saved: Cropped_Tumors\benign\tumor_789_mask0.png
Saved: Cropped_Tumors\benign\tumor_789_mask1.png
Saved: Cropped_Tumors\benign\tumor_792_mask0.png
Saved: Cropped_Tumors\benign\tumor_794_mask0.png
Saved: Cropped_Tumors\benign\tumor_796_mask0.png
Saved: Cropped_Tumors\benign\tumor_798_mask0.png
Saved: Cropped_Tumors\benign\tumor_800_mask0.png
Saved: Cropped_Tumors\benign\tumor_802_mask0.png
Saved: Cropped_Tumors\benign\tumor_804_mask0.png
Saved: Cropped_Tumors\benign\tumor_806_mask0.png


 60%|█████▉    | 812/1357 [00:07<00:07, 76.53it/s]

Saved: Cropped_Tumors\benign\tumor_808_mask0.png
Saved: Cropped_Tumors\benign\tumor_810_mask0.png
Saved: Cropped_Tumors\benign\tumor_812_mask0.png
Saved: Cropped_Tumors\benign\tumor_814_mask0.png
Saved: Cropped_Tumors\benign\tumor_816_mask0.png
Saved: Cropped_Tumors\benign\tumor_818_mask0.png


 61%|██████    | 828/1357 [00:08<00:07, 70.90it/s]

Saved: Cropped_Tumors\benign\tumor_820_mask0.png
Saved: Cropped_Tumors\benign\tumor_822_mask0.png
Saved: Cropped_Tumors\benign\tumor_824_mask0.png
Saved: Cropped_Tumors\benign\tumor_826_mask0.png
Saved: Cropped_Tumors\benign\tumor_828_mask0.png
Saved: Cropped_Tumors\benign\tumor_830_mask0.png
Saved: Cropped_Tumors\benign\tumor_832_mask0.png
Saved: Cropped_Tumors\benign\tumor_834_mask0.png
Saved: Cropped_Tumors\benign\tumor_836_mask0.png
Saved: Cropped_Tumors\benign\tumor_838_mask0.png


 63%|██████▎   | 851/1357 [00:08<00:07, 69.16it/s]

Saved: Cropped_Tumors\benign\tumor_840_mask0.png
Saved: Cropped_Tumors\benign\tumor_842_mask0.png
Saved: Cropped_Tumors\benign\tumor_844_mask0.png
Saved: Cropped_Tumors\benign\tumor_846_mask0.png
Saved: Cropped_Tumors\benign\tumor_846_mask1.png
Saved: Cropped_Tumors\benign\tumor_849_mask0.png
Saved: Cropped_Tumors\benign\tumor_851_mask0.png


 63%|██████▎   | 861/1357 [00:08<00:07, 67.86it/s]

Saved: Cropped_Tumors\benign\tumor_853_mask0.png
Saved: Cropped_Tumors\benign\tumor_855_mask0.png
Saved: Cropped_Tumors\benign\tumor_857_mask0.png
Saved: Cropped_Tumors\benign\tumor_859_mask0.png
Saved: Cropped_Tumors\benign\tumor_861_mask0.png
Saved: Cropped_Tumors\benign\tumor_863_mask0.png
Saved: Cropped_Tumors\benign\tumor_865_mask0.png
Saved: Cropped_Tumors\benign\tumor_867_mask0.png


 65%|██████▍   | 878/1357 [00:09<00:07, 65.72it/s]

Saved: Cropped_Tumors\benign\tumor_869_mask0.png
Saved: Cropped_Tumors\benign\tumor_871_mask0.png
Saved: Cropped_Tumors\benign\tumor_873_mask0.png
Saved: Cropped_Tumors\benign\tumor_875_mask0.png
Saved: Cropped_Tumors\benign\tumor_877_mask0.png
Saved: Cropped_Tumors\benign\tumor_879_mask0.png
Saved: Cropped_Tumors\benign\tumor_881_mask0.png


 66%|██████▌   | 894/1357 [00:09<00:07, 65.84it/s]

Saved: Cropped_Tumors\benign\tumor_883_mask0.png
Saved: Cropped_Tumors\benign\tumor_885_mask0.png
Saved: Cropped_Tumors\benign\tumor_887_mask0.png
Saved: Cropped_Tumors\benign\tumor_889_mask0.png
Saved: Cropped_Tumors\benign\tumor_891_mask0.png
Saved: Cropped_Tumors\benign\tumor_893_mask0.png
Saved: Cropped_Tumors\benign\tumor_895_mask0.png
Saved: Cropped_Tumors\benign\tumor_897_mask0.png


 67%|██████▋   | 910/1357 [00:09<00:06, 65.58it/s]

Saved: Cropped_Tumors\benign\tumor_897_mask1.png
Saved: Cropped_Tumors\benign\tumor_900_mask0.png
Saved: Cropped_Tumors\benign\tumor_902_mask0.png
Saved: Cropped_Tumors\benign\tumor_904_mask0.png
Saved: Cropped_Tumors\benign\tumor_906_mask0.png
Saved: Cropped_Tumors\benign\tumor_906_mask1.png
Saved: Cropped_Tumors\benign\tumor_909_mask0.png


 68%|██████▊   | 924/1357 [00:09<00:06, 63.62it/s]

Saved: Cropped_Tumors\benign\tumor_911_mask0.png
Saved: Cropped_Tumors\benign\tumor_913_mask0.png
Saved: Cropped_Tumors\benign\tumor_915_mask0.png
Saved: Cropped_Tumors\benign\tumor_917_mask0.png
Saved: Cropped_Tumors\benign\tumor_919_mask0.png
Saved: Cropped_Tumors\benign\tumor_921_mask0.png
Saved: Cropped_Tumors\benign\tumor_923_mask0.png


 69%|██████▊   | 932/1357 [00:09<00:06, 63.11it/s]

Saved: Cropped_Tumors\benign\tumor_925_mask0.png
Saved: Cropped_Tumors\benign\tumor_927_mask0.png
Saved: Cropped_Tumors\benign\tumor_929_mask0.png
Saved: Cropped_Tumors\benign\tumor_931_mask0.png
Saved: Cropped_Tumors\benign\tumor_933_mask0.png
Saved: Cropped_Tumors\benign\tumor_935_mask0.png


 70%|██████▉   | 946/1357 [00:10<00:06, 62.01it/s]

Saved: Cropped_Tumors\benign\tumor_937_mask0.png
Saved: Cropped_Tumors\benign\tumor_939_mask0.png
Saved: Cropped_Tumors\benign\tumor_941_mask0.png
Saved: Cropped_Tumors\benign\tumor_943_mask0.png
Saved: Cropped_Tumors\benign\tumor_945_mask0.png
Saved: Cropped_Tumors\benign\tumor_947_mask0.png
Saved: Cropped_Tumors\benign\tumor_949_mask0.png


 71%|███████   | 961/1357 [00:10<00:06, 62.99it/s]

Saved: Cropped_Tumors\benign\tumor_951_mask0.png
Saved: Cropped_Tumors\benign\tumor_953_mask0.png
Saved: Cropped_Tumors\benign\tumor_955_mask0.png
Saved: Cropped_Tumors\benign\tumor_957_mask0.png
Saved: Cropped_Tumors\benign\tumor_957_mask1.png
Saved: Cropped_Tumors\benign\tumor_960_mask0.png
Saved: Cropped_Tumors\benign\tumor_962_mask0.png


 72%|███████▏  | 975/1357 [00:10<00:06, 60.25it/s]

Saved: Cropped_Tumors\benign\tumor_964_mask0.png
Saved: Cropped_Tumors\benign\tumor_966_mask0.png
Saved: Cropped_Tumors\benign\tumor_968_mask0.png
Saved: Cropped_Tumors\benign\tumor_970_mask0.png
Saved: Cropped_Tumors\benign\tumor_972_mask0.png
Saved: Cropped_Tumors\benign\tumor_974_mask0.png
Saved: Cropped_Tumors\benign\tumor_976_mask0.png


 73%|███████▎  | 991/1357 [00:10<00:06, 58.72it/s]

Saved: Cropped_Tumors\benign\tumor_978_mask0.png
Saved: Cropped_Tumors\benign\tumor_978_mask1.png
Saved: Cropped_Tumors\benign\tumor_981_mask0.png
Saved: Cropped_Tumors\benign\tumor_981_mask1.png
Saved: Cropped_Tumors\benign\tumor_984_mask0.png
Saved: Cropped_Tumors\benign\tumor_986_mask0.png
Saved: Cropped_Tumors\benign\tumor_988_mask0.png
Saved: Cropped_Tumors\benign\tumor_990_mask0.png
Saved: Cropped_Tumors\benign\tumor_990_mask1.png


 74%|███████▍  | 1003/1357 [00:11<00:06, 55.00it/s]

Saved: Cropped_Tumors\benign\tumor_993_mask0.png
Saved: Cropped_Tumors\malignant\tumor_995_mask0.png
Saved: Cropped_Tumors\malignant\tumor_997_mask0.png
Saved: Cropped_Tumors\malignant\tumor_999_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1001_mask0.png


 75%|███████▍  | 1015/1357 [00:11<00:06, 52.18it/s]

Saved: Cropped_Tumors\malignant\tumor_1003_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1005_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1007_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1009_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1011_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1013_mask0.png


 75%|███████▌  | 1022/1357 [00:11<00:06, 54.07it/s]

Saved: Cropped_Tumors\malignant\tumor_1015_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1017_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1019_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1021_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1023_mask0.png


 76%|███████▌  | 1034/1357 [00:11<00:06, 50.32it/s]

Saved: Cropped_Tumors\malignant\tumor_1025_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1027_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1029_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1031_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1033_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1035_mask0.png


 77%|███████▋  | 1046/1357 [00:12<00:06, 50.29it/s]

Saved: Cropped_Tumors\malignant\tumor_1037_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1039_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1041_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1043_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1045_mask0.png


 78%|███████▊  | 1058/1357 [00:12<00:05, 53.36it/s]

Saved: Cropped_Tumors\malignant\tumor_1047_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1049_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1051_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1053_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1055_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1057_mask0.png


 78%|███████▊  | 1064/1357 [00:12<00:05, 52.86it/s]

Saved: Cropped_Tumors\malignant\tumor_1059_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1061_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1063_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1065_mask0.png


 79%|███████▉  | 1070/1357 [00:12<00:06, 43.91it/s]

Saved: Cropped_Tumors\malignant\tumor_1067_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1069_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1071_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1073_mask0.png


 80%|███████▉  | 1082/1357 [00:12<00:06, 44.56it/s]

Saved: Cropped_Tumors\malignant\tumor_1075_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1077_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1079_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1081_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1083_mask0.png


 80%|████████  | 1087/1357 [00:12<00:05, 45.48it/s]

Saved: Cropped_Tumors\malignant\tumor_1085_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1087_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1089_mask0.png


 81%|████████  | 1098/1357 [00:13<00:06, 39.25it/s]

Saved: Cropped_Tumors\malignant\tumor_1091_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1093_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1095_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1097_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1099_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1101_mask0.png


 82%|████████▏ | 1110/1357 [00:13<00:05, 44.00it/s]

Saved: Cropped_Tumors\malignant\tumor_1103_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1105_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1107_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1109_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1111_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1113_mask0.png


 83%|████████▎ | 1124/1357 [00:13<00:04, 50.78it/s]

Saved: Cropped_Tumors\malignant\tumor_1115_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1117_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1119_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1121_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1123_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1125_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1127_mask0.png


 84%|████████▎ | 1136/1357 [00:14<00:04, 47.34it/s]

Saved: Cropped_Tumors\malignant\tumor_1129_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1131_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1133_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1135_mask0.png


 85%|████████▍ | 1148/1357 [00:14<00:04, 49.24it/s]

Saved: Cropped_Tumors\malignant\tumor_1137_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1139_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1141_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1143_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1145_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1147_mask0.png


 85%|████████▌ | 1156/1357 [00:14<00:03, 52.34it/s]

Saved: Cropped_Tumors\malignant\tumor_1149_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1151_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1153_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1155_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1157_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1159_mask0.png


 86%|████████▌ | 1168/1357 [00:14<00:03, 49.67it/s]

Saved: Cropped_Tumors\malignant\tumor_1161_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1163_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1165_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1167_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1169_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1171_mask0.png


 87%|████████▋ | 1180/1357 [00:14<00:03, 50.59it/s]

Saved: Cropped_Tumors\malignant\tumor_1173_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1175_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1177_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1179_mask0.png


 88%|████████▊ | 1192/1357 [00:15<00:03, 45.52it/s]

Saved: Cropped_Tumors\malignant\tumor_1181_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1183_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1185_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1187_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1189_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1191_mask0.png


 89%|████████▊ | 1204/1357 [00:15<00:03, 47.22it/s]

Saved: Cropped_Tumors\malignant\tumor_1193_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1195_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1197_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1199_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1201_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1203_mask0.png


 90%|████████▉ | 1215/1357 [00:15<00:03, 45.46it/s]

Saved: Cropped_Tumors\malignant\tumor_1205_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1207_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1209_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1211_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1213_mask0.png


 90%|█████████ | 1226/1357 [00:15<00:02, 46.66it/s]

Saved: Cropped_Tumors\malignant\tumor_1215_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1217_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1219_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1221_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1223_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1225_mask0.png


 91%|█████████ | 1238/1357 [00:16<00:02, 49.36it/s]

Saved: Cropped_Tumors\malignant\tumor_1227_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1229_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1231_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1233_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1235_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1237_mask0.png


 92%|█████████▏| 1250/1357 [00:16<00:02, 52.70it/s]

Saved: Cropped_Tumors\malignant\tumor_1239_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1241_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1243_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1245_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1247_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1249_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1251_mask0.png


 93%|█████████▎| 1262/1357 [00:16<00:01, 49.36it/s]

Saved: Cropped_Tumors\malignant\tumor_1253_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1255_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1257_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1259_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1261_mask0.png


 94%|█████████▍| 1274/1357 [00:16<00:01, 50.28it/s]

Saved: Cropped_Tumors\malignant\tumor_1263_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1265_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1267_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1269_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1271_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1273_mask0.png


 95%|█████████▍| 1286/1357 [00:17<00:01, 50.65it/s]

Saved: Cropped_Tumors\malignant\tumor_1275_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1277_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1279_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1281_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1283_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1285_mask0.png


 95%|█████████▌| 1292/1357 [00:17<00:01, 46.27it/s]

Saved: Cropped_Tumors\malignant\tumor_1287_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1289_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1291_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1293_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1295_mask0.png


 96%|█████████▌| 1304/1357 [00:17<00:01, 51.51it/s]

Saved: Cropped_Tumors\malignant\tumor_1297_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1299_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1301_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1303_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1305_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1307_mask0.png


 97%|█████████▋| 1316/1357 [00:17<00:00, 49.52it/s]

Saved: Cropped_Tumors\malignant\tumor_1309_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1311_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1313_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1315_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1317_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1319_mask0.png


 98%|█████████▊| 1328/1357 [00:17<00:00, 49.61it/s]

Saved: Cropped_Tumors\malignant\tumor_1321_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1323_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1325_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1327_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1329_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1331_mask0.png


 99%|█████████▊| 1340/1357 [00:18<00:00, 47.32it/s]

Saved: Cropped_Tumors\malignant\tumor_1333_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1335_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1337_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1339_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1341_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1343_mask0.png


100%|██████████| 1357/1357 [00:18<00:00, 73.20it/s]

Saved: Cropped_Tumors\malignant\tumor_1345_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1347_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1349_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1351_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1353_mask0.png
Saved: Cropped_Tumors\malignant\tumor_1355_mask0.png



  1%|          | 14/1357 [00:00<00:21, 61.88it/s]

Saved: Cropped_Tumors\normal\normal_0.png
Saved: Cropped_Tumors\normal\normal_2.png
Saved: Cropped_Tumors\normal\normal_4.png
Saved: Cropped_Tumors\normal\normal_6.png
Saved: Cropped_Tumors\normal\normal_8.png
Saved: Cropped_Tumors\normal\normal_10.png
Saved: Cropped_Tumors\normal\normal_12.png


  2%|▏         | 29/1357 [00:00<00:20, 65.69it/s]

Saved: Cropped_Tumors\normal\normal_14.png
Saved: Cropped_Tumors\normal\normal_16.png
Saved: Cropped_Tumors\normal\normal_18.png
Saved: Cropped_Tumors\normal\normal_20.png
Saved: Cropped_Tumors\normal\normal_22.png
Saved: Cropped_Tumors\normal\normal_24.png
Saved: Cropped_Tumors\normal\normal_26.png
Saved: Cropped_Tumors\normal\normal_28.png


  3%|▎         | 36/1357 [00:00<00:21, 62.84it/s]

Saved: Cropped_Tumors\normal\normal_30.png
Saved: Cropped_Tumors\normal\normal_32.png
Saved: Cropped_Tumors\normal\normal_34.png
Saved: Cropped_Tumors\normal\normal_36.png
Saved: Cropped_Tumors\normal\normal_38.png
Saved: Cropped_Tumors\normal\normal_40.png


  4%|▍         | 51/1357 [00:00<00:22, 59.21it/s]

Saved: Cropped_Tumors\normal\normal_42.png
Saved: Cropped_Tumors\normal\normal_44.png
Saved: Cropped_Tumors\normal\normal_46.png
Saved: Cropped_Tumors\normal\normal_48.png
Saved: Cropped_Tumors\normal\normal_50.png
Saved: Cropped_Tumors\normal\normal_52.png
Saved: Cropped_Tumors\normal\normal_54.png


  5%|▍         | 65/1357 [00:01<00:21, 60.19it/s]

Saved: Cropped_Tumors\normal\normal_56.png
Saved: Cropped_Tumors\normal\normal_58.png
Saved: Cropped_Tumors\normal\normal_60.png
Saved: Cropped_Tumors\normal\normal_62.png
Saved: Cropped_Tumors\normal\normal_64.png
Saved: Cropped_Tumors\normal\normal_66.png
Saved: Cropped_Tumors\normal\normal_68.png


  6%|▌         | 81/1357 [00:01<00:20, 61.56it/s]

Saved: Cropped_Tumors\normal\normal_70.png
Saved: Cropped_Tumors\normal\normal_72.png
Saved: Cropped_Tumors\normal\normal_74.png
Saved: Cropped_Tumors\normal\normal_76.png
Saved: Cropped_Tumors\normal\normal_78.png
Saved: Cropped_Tumors\normal\normal_80.png
Saved: Cropped_Tumors\normal\normal_82.png
Saved: Cropped_Tumors\normal\normal_84.png


  7%|▋         | 96/1357 [00:01<00:18, 66.79it/s]

Saved: Cropped_Tumors\normal\normal_86.png
Saved: Cropped_Tumors\normal\normal_88.png
Saved: Cropped_Tumors\normal\normal_90.png
Saved: Cropped_Tumors\normal\normal_92.png
Saved: Cropped_Tumors\normal\normal_94.png
Saved: Cropped_Tumors\normal\normal_96.png


  8%|▊         | 110/1357 [00:01<00:20, 62.00it/s]

Saved: Cropped_Tumors\normal\normal_98.png
Saved: Cropped_Tumors\normal\normal_100.png
Saved: Cropped_Tumors\normal\normal_102.png
Saved: Cropped_Tumors\normal\normal_104.png
Saved: Cropped_Tumors\normal\normal_106.png
Saved: Cropped_Tumors\normal\normal_108.png
Saved: Cropped_Tumors\normal\normal_110.png


  9%|▉         | 127/1357 [00:02<00:19, 62.58it/s]

Saved: Cropped_Tumors\normal\normal_112.png
Saved: Cropped_Tumors\normal\normal_114.png
Saved: Cropped_Tumors\normal\normal_116.png
Saved: Cropped_Tumors\normal\normal_118.png
Saved: Cropped_Tumors\normal\normal_120.png
Saved: Cropped_Tumors\normal\normal_122.png
Saved: Cropped_Tumors\normal\normal_124.png
Saved: Cropped_Tumors\normal\normal_126.png


 10%|▉         | 134/1357 [00:02<00:21, 57.62it/s]

Saved: Cropped_Tumors\normal\normal_128.png
Saved: Cropped_Tumors\normal\normal_130.png
Saved: Cropped_Tumors\normal\normal_132.png
Saved: Cropped_Tumors\normal\normal_134.png
Saved: Cropped_Tumors\normal\normal_136.png
Saved: Cropped_Tumors\normal\normal_138.png


 11%|█         | 147/1357 [00:02<00:21, 57.20it/s]

Saved: Cropped_Tumors\normal\normal_140.png
Saved: Cropped_Tumors\normal\normal_142.png
Saved: Cropped_Tumors\normal\normal_144.png
Saved: Cropped_Tumors\normal\normal_146.png
Saved: Cropped_Tumors\normal\normal_148.png
Saved: Cropped_Tumors\normal\normal_150.png
Saved: Cropped_Tumors\normal\normal_152.png
Saved: Cropped_Tumors\normal\normal_154.png


 12%|█▏        | 165/1357 [00:02<00:17, 67.95it/s]

Saved: Cropped_Tumors\normal\normal_156.png
Saved: Cropped_Tumors\normal\normal_158.png
Saved: Cropped_Tumors\normal\normal_160.png
Saved: Cropped_Tumors\normal\normal_162.png
Saved: Cropped_Tumors\normal\normal_164.png
Saved: Cropped_Tumors\normal\normal_166.png
Saved: Cropped_Tumors\normal\normal_168.png


 13%|█▎        | 181/1357 [00:02<00:17, 68.60it/s]

Saved: Cropped_Tumors\normal\normal_170.png
Saved: Cropped_Tumors\normal\normal_172.png
Saved: Cropped_Tumors\normal\normal_174.png
Saved: Cropped_Tumors\normal\normal_176.png
Saved: Cropped_Tumors\normal\normal_178.png
Saved: Cropped_Tumors\normal\normal_180.png
Saved: Cropped_Tumors\normal\normal_182.png


 15%|█▍        | 197/1357 [00:03<00:16, 69.88it/s]

Saved: Cropped_Tumors\normal\normal_184.png
Saved: Cropped_Tumors\normal\normal_186.png
Saved: Cropped_Tumors\normal\normal_188.png
Saved: Cropped_Tumors\normal\normal_190.png
Saved: Cropped_Tumors\normal\normal_192.png
Saved: Cropped_Tumors\normal\normal_194.png
Saved: Cropped_Tumors\normal\normal_196.png
Saved: Cropped_Tumors\normal\normal_198.png


 16%|█▌        | 213/1357 [00:03<00:17, 65.86it/s]

Saved: Cropped_Tumors\normal\normal_200.png
Saved: Cropped_Tumors\normal\normal_202.png
Saved: Cropped_Tumors\normal\normal_204.png
Saved: Cropped_Tumors\normal\normal_206.png
Saved: Cropped_Tumors\normal\normal_208.png
Saved: Cropped_Tumors\normal\normal_210.png
Saved: Cropped_Tumors\normal\normal_212.png


 16%|█▋        | 223/1357 [00:03<00:16, 69.41it/s]

Saved: Cropped_Tumors\normal\normal_214.png
Saved: Cropped_Tumors\normal\normal_216.png
Saved: Cropped_Tumors\normal\normal_218.png
Saved: Cropped_Tumors\normal\normal_220.png
Saved: Cropped_Tumors\normal\normal_222.png
Saved: Cropped_Tumors\normal\normal_224.png
Saved: Cropped_Tumors\normal\normal_226.png
Saved: Cropped_Tumors\normal\normal_228.png


 18%|█▊        | 239/1357 [00:03<00:15, 70.64it/s]

Saved: Cropped_Tumors\normal\normal_230.png
Saved: Cropped_Tumors\normal\normal_232.png
Saved: Cropped_Tumors\normal\normal_234.png
Saved: Cropped_Tumors\normal\normal_236.png
Saved: Cropped_Tumors\normal\normal_238.png
Saved: Cropped_Tumors\normal\normal_240.png
Saved: Cropped_Tumors\normal\normal_242.png
Saved: Cropped_Tumors\normal\normal_244.png


 19%|█▉        | 257/1357 [00:03<00:15, 71.75it/s]

Saved: Cropped_Tumors\normal\normal_246.png
Saved: Cropped_Tumors\normal\normal_248.png
Saved: Cropped_Tumors\normal\normal_250.png
Saved: Cropped_Tumors\normal\normal_252.png
Saved: Cropped_Tumors\normal\normal_254.png
Saved: Cropped_Tumors\normal\normal_256.png
Saved: Cropped_Tumors\normal\normal_258.png
Saved: Cropped_Tumors\normal\normal_260.png


100%|██████████| 1357/1357 [00:04<00:00, 329.50it/s]

Saved: Cropped_Tumors\normal\normal_262.png
Saved: Cropped_Tumors\normal\normal_264.png
Processing complete. Cropped tumor images and normal images are saved in: Cropped_Tumors


We now generates 4 random crops per normal image in order to increase the number of images we feed our CNN.  

In [44]:
import random

# Define output directory for cropped normal images
output_dir = "Cropped_Tumors"
normal_dir = os.path.join(output_dir, "normal")

os.makedirs(normal_dir, exist_ok=True)

# Random crop parameters
NUM_CROPS = 4  # Number of crops per normal image
CROP_SIZE = 128  # Size of each cropped region (128x128 pixels)

# Iterate through the DataFrame and process only normal ultrasound images
for index, row in tqdm(df.iterrows(), total=len(df)):
    image_path = row["image_path"]
    label = row["label"]  # 0 = normal

    # Skip non-normal images and all mask images
    if label != 0 or "_mask" in image_path:
        continue  

    # Load the normal ultrasound image
    image = cv2.imread(image_path)
    if image is None:
        print(f"Warning: Could not load {image_path}")
        continue
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image = cv2.resize(image, (1024, 1024))  # Resize for consistency

    # Generate random crops
    for crop_index in range(NUM_CROPS):
        # Randomly select a top-left corner for the crop (within image bounds)
        x = random.randint(0, 1024 - CROP_SIZE)
        y = random.randint(0, 1024 - CROP_SIZE)

        # Extract the crop
        crop = image[y:y + CROP_SIZE, x:x + CROP_SIZE]

        # Save the cropped image
        crop_filename = os.path.join(normal_dir, f"normal_{index}_crop{crop_index}.png")
        cv2.imwrite(crop_filename, cv2.cvtColor(crop, cv2.COLOR_RGB2BGR))

        print(f"Saved: {crop_filename}")

print("Processing complete. Randomly cropped normal images are saved in:", normal_dir)


  1%|          | 7/1357 [00:00<00:22, 60.66it/s]

Saved: Cropped_Tumors\normal\normal_0_crop0.png
Saved: Cropped_Tumors\normal\normal_0_crop1.png
Saved: Cropped_Tumors\normal\normal_0_crop2.png
Saved: Cropped_Tumors\normal\normal_0_crop3.png
Saved: Cropped_Tumors\normal\normal_2_crop0.png
Saved: Cropped_Tumors\normal\normal_2_crop1.png
Saved: Cropped_Tumors\normal\normal_2_crop2.png
Saved: Cropped_Tumors\normal\normal_2_crop3.png
Saved: Cropped_Tumors\normal\normal_4_crop0.png
Saved: Cropped_Tumors\normal\normal_4_crop1.png
Saved: Cropped_Tumors\normal\normal_4_crop2.png
Saved: Cropped_Tumors\normal\normal_4_crop3.png
Saved: Cropped_Tumors\normal\normal_6_crop0.png
Saved: Cropped_Tumors\normal\normal_6_crop1.png
Saved: Cropped_Tumors\normal\normal_6_crop2.png
Saved: Cropped_Tumors\normal\normal_6_crop3.png
Saved: Cropped_Tumors\normal\normal_8_crop0.png
Saved: Cropped_Tumors\normal\normal_8_crop1.png
Saved: Cropped_Tumors\normal\normal_8_crop2.png
Saved: Cropped_Tumors\normal\normal_8_crop3.png
Saved: Cropped_Tumors\normal\normal_10_c

  2%|▏         | 23/1357 [00:00<00:20, 65.77it/s]

Saved: Cropped_Tumors\normal\normal_12_crop1.png
Saved: Cropped_Tumors\normal\normal_12_crop2.png
Saved: Cropped_Tumors\normal\normal_12_crop3.png
Saved: Cropped_Tumors\normal\normal_14_crop0.png
Saved: Cropped_Tumors\normal\normal_14_crop1.png
Saved: Cropped_Tumors\normal\normal_14_crop2.png
Saved: Cropped_Tumors\normal\normal_14_crop3.png
Saved: Cropped_Tumors\normal\normal_16_crop0.png
Saved: Cropped_Tumors\normal\normal_16_crop1.png
Saved: Cropped_Tumors\normal\normal_16_crop2.png
Saved: Cropped_Tumors\normal\normal_16_crop3.png
Saved: Cropped_Tumors\normal\normal_18_crop0.png
Saved: Cropped_Tumors\normal\normal_18_crop1.png
Saved: Cropped_Tumors\normal\normal_18_crop2.png
Saved: Cropped_Tumors\normal\normal_18_crop3.png
Saved: Cropped_Tumors\normal\normal_20_crop0.png
Saved: Cropped_Tumors\normal\normal_20_crop1.png
Saved: Cropped_Tumors\normal\normal_20_crop2.png
Saved: Cropped_Tumors\normal\normal_20_crop3.png
Saved: Cropped_Tumors\normal\normal_22_crop0.png
Saved: Cropped_Tumor

  3%|▎         | 38/1357 [00:00<00:20, 65.37it/s]

Saved: Cropped_Tumors\normal\normal_28_crop0.png
Saved: Cropped_Tumors\normal\normal_28_crop1.png
Saved: Cropped_Tumors\normal\normal_28_crop2.png
Saved: Cropped_Tumors\normal\normal_28_crop3.png
Saved: Cropped_Tumors\normal\normal_30_crop0.png
Saved: Cropped_Tumors\normal\normal_30_crop1.png
Saved: Cropped_Tumors\normal\normal_30_crop2.png
Saved: Cropped_Tumors\normal\normal_30_crop3.png
Saved: Cropped_Tumors\normal\normal_32_crop0.png
Saved: Cropped_Tumors\normal\normal_32_crop1.png
Saved: Cropped_Tumors\normal\normal_32_crop2.png
Saved: Cropped_Tumors\normal\normal_32_crop3.png
Saved: Cropped_Tumors\normal\normal_34_crop0.png
Saved: Cropped_Tumors\normal\normal_34_crop1.png
Saved: Cropped_Tumors\normal\normal_34_crop2.png
Saved: Cropped_Tumors\normal\normal_34_crop3.png
Saved: Cropped_Tumors\normal\normal_36_crop0.png
Saved: Cropped_Tumors\normal\normal_36_crop1.png
Saved: Cropped_Tumors\normal\normal_36_crop2.png
Saved: Cropped_Tumors\normal\normal_36_crop3.png
Saved: Cropped_Tumor

  4%|▍         | 52/1357 [00:00<00:20, 62.77it/s]

Saved: Cropped_Tumors\normal\normal_40_crop2.png
Saved: Cropped_Tumors\normal\normal_40_crop3.png
Saved: Cropped_Tumors\normal\normal_42_crop0.png
Saved: Cropped_Tumors\normal\normal_42_crop1.png
Saved: Cropped_Tumors\normal\normal_42_crop2.png
Saved: Cropped_Tumors\normal\normal_42_crop3.png
Saved: Cropped_Tumors\normal\normal_44_crop0.png
Saved: Cropped_Tumors\normal\normal_44_crop1.png
Saved: Cropped_Tumors\normal\normal_44_crop2.png
Saved: Cropped_Tumors\normal\normal_44_crop3.png
Saved: Cropped_Tumors\normal\normal_46_crop0.png
Saved: Cropped_Tumors\normal\normal_46_crop1.png
Saved: Cropped_Tumors\normal\normal_46_crop2.png
Saved: Cropped_Tumors\normal\normal_46_crop3.png
Saved: Cropped_Tumors\normal\normal_48_crop0.png
Saved: Cropped_Tumors\normal\normal_48_crop1.png
Saved: Cropped_Tumors\normal\normal_48_crop2.png
Saved: Cropped_Tumors\normal\normal_48_crop3.png
Saved: Cropped_Tumors\normal\normal_50_crop0.png
Saved: Cropped_Tumors\normal\normal_50_crop1.png
Saved: Cropped_Tumor

  5%|▍         | 67/1357 [00:01<00:20, 61.56it/s]

Saved: Cropped_Tumors\normal\normal_54_crop0.png
Saved: Cropped_Tumors\normal\normal_54_crop1.png
Saved: Cropped_Tumors\normal\normal_54_crop2.png
Saved: Cropped_Tumors\normal\normal_54_crop3.png
Saved: Cropped_Tumors\normal\normal_56_crop0.png
Saved: Cropped_Tumors\normal\normal_56_crop1.png
Saved: Cropped_Tumors\normal\normal_56_crop2.png
Saved: Cropped_Tumors\normal\normal_56_crop3.png
Saved: Cropped_Tumors\normal\normal_58_crop0.png
Saved: Cropped_Tumors\normal\normal_58_crop1.png
Saved: Cropped_Tumors\normal\normal_58_crop2.png
Saved: Cropped_Tumors\normal\normal_58_crop3.png
Saved: Cropped_Tumors\normal\normal_60_crop0.png
Saved: Cropped_Tumors\normal\normal_60_crop1.png
Saved: Cropped_Tumors\normal\normal_60_crop2.png
Saved: Cropped_Tumors\normal\normal_60_crop3.png
Saved: Cropped_Tumors\normal\normal_62_crop0.png
Saved: Cropped_Tumors\normal\normal_62_crop1.png
Saved: Cropped_Tumors\normal\normal_62_crop2.png
Saved: Cropped_Tumors\normal\normal_62_crop3.png
Saved: Cropped_Tumor

  6%|▌         | 75/1357 [00:01<00:20, 61.70it/s]

Saved: Cropped_Tumors\normal\normal_68_crop0.png
Saved: Cropped_Tumors\normal\normal_68_crop1.png
Saved: Cropped_Tumors\normal\normal_68_crop2.png
Saved: Cropped_Tumors\normal\normal_68_crop3.png
Saved: Cropped_Tumors\normal\normal_70_crop0.png
Saved: Cropped_Tumors\normal\normal_70_crop1.png
Saved: Cropped_Tumors\normal\normal_70_crop2.png
Saved: Cropped_Tumors\normal\normal_70_crop3.png
Saved: Cropped_Tumors\normal\normal_72_crop0.png
Saved: Cropped_Tumors\normal\normal_72_crop1.png
Saved: Cropped_Tumors\normal\normal_72_crop2.png
Saved: Cropped_Tumors\normal\normal_72_crop3.png
Saved: Cropped_Tumors\normal\normal_74_crop0.png
Saved: Cropped_Tumors\normal\normal_74_crop1.png
Saved: Cropped_Tumors\normal\normal_74_crop2.png
Saved: Cropped_Tumors\normal\normal_74_crop3.png
Saved: Cropped_Tumors\normal\normal_76_crop0.png
Saved: Cropped_Tumors\normal\normal_76_crop1.png
Saved: Cropped_Tumors\normal\normal_76_crop2.png
Saved: Cropped_Tumors\normal\normal_76_crop3.png
Saved: Cropped_Tumor

  7%|▋         | 92/1357 [00:01<00:18, 68.17it/s]

Saved: Cropped_Tumors\normal\normal_84_crop0.png
Saved: Cropped_Tumors\normal\normal_84_crop1.png
Saved: Cropped_Tumors\normal\normal_84_crop2.png
Saved: Cropped_Tumors\normal\normal_84_crop3.png
Saved: Cropped_Tumors\normal\normal_86_crop0.png
Saved: Cropped_Tumors\normal\normal_86_crop1.png
Saved: Cropped_Tumors\normal\normal_86_crop2.png
Saved: Cropped_Tumors\normal\normal_86_crop3.png
Saved: Cropped_Tumors\normal\normal_88_crop0.png
Saved: Cropped_Tumors\normal\normal_88_crop1.png
Saved: Cropped_Tumors\normal\normal_88_crop2.png
Saved: Cropped_Tumors\normal\normal_88_crop3.png
Saved: Cropped_Tumors\normal\normal_90_crop0.png
Saved: Cropped_Tumors\normal\normal_90_crop1.png
Saved: Cropped_Tumors\normal\normal_90_crop2.png
Saved: Cropped_Tumors\normal\normal_90_crop3.png
Saved: Cropped_Tumors\normal\normal_92_crop0.png
Saved: Cropped_Tumors\normal\normal_92_crop1.png
Saved: Cropped_Tumors\normal\normal_92_crop2.png
Saved: Cropped_Tumors\normal\normal_92_crop3.png
Saved: Cropped_Tumor

  8%|▊         | 107/1357 [00:01<00:19, 64.83it/s]

Saved: Cropped_Tumors\normal\normal_98_crop0.png
Saved: Cropped_Tumors\normal\normal_98_crop1.png
Saved: Cropped_Tumors\normal\normal_98_crop2.png
Saved: Cropped_Tumors\normal\normal_98_crop3.png
Saved: Cropped_Tumors\normal\normal_100_crop0.png
Saved: Cropped_Tumors\normal\normal_100_crop1.png
Saved: Cropped_Tumors\normal\normal_100_crop2.png
Saved: Cropped_Tumors\normal\normal_100_crop3.png
Saved: Cropped_Tumors\normal\normal_102_crop0.png
Saved: Cropped_Tumors\normal\normal_102_crop1.png
Saved: Cropped_Tumors\normal\normal_102_crop2.png
Saved: Cropped_Tumors\normal\normal_102_crop3.png
Saved: Cropped_Tumors\normal\normal_104_crop0.png
Saved: Cropped_Tumors\normal\normal_104_crop1.png
Saved: Cropped_Tumors\normal\normal_104_crop2.png
Saved: Cropped_Tumors\normal\normal_104_crop3.png
Saved: Cropped_Tumors\normal\normal_106_crop0.png
Saved: Cropped_Tumors\normal\normal_106_crop1.png
Saved: Cropped_Tumors\normal\normal_106_crop2.png
Saved: Cropped_Tumors\normal\normal_106_crop3.png
Save

  9%|▉         | 123/1357 [00:01<00:17, 69.13it/s]

Saved: Cropped_Tumors\normal\normal_112_crop0.png
Saved: Cropped_Tumors\normal\normal_112_crop1.png
Saved: Cropped_Tumors\normal\normal_112_crop2.png
Saved: Cropped_Tumors\normal\normal_112_crop3.png
Saved: Cropped_Tumors\normal\normal_114_crop0.png
Saved: Cropped_Tumors\normal\normal_114_crop1.png
Saved: Cropped_Tumors\normal\normal_114_crop2.png
Saved: Cropped_Tumors\normal\normal_114_crop3.png
Saved: Cropped_Tumors\normal\normal_116_crop0.png
Saved: Cropped_Tumors\normal\normal_116_crop1.png
Saved: Cropped_Tumors\normal\normal_116_crop2.png
Saved: Cropped_Tumors\normal\normal_116_crop3.png
Saved: Cropped_Tumors\normal\normal_118_crop0.png
Saved: Cropped_Tumors\normal\normal_118_crop1.png
Saved: Cropped_Tumors\normal\normal_118_crop2.png
Saved: Cropped_Tumors\normal\normal_118_crop3.png
Saved: Cropped_Tumors\normal\normal_120_crop0.png
Saved: Cropped_Tumors\normal\normal_120_crop1.png
Saved: Cropped_Tumors\normal\normal_120_crop2.png
Saved: Cropped_Tumors\normal\normal_120_crop3.png


 10%|█         | 137/1357 [00:02<00:18, 65.52it/s]

Saved: Cropped_Tumors\normal\normal_126_crop3.png
Saved: Cropped_Tumors\normal\normal_128_crop0.png
Saved: Cropped_Tumors\normal\normal_128_crop1.png
Saved: Cropped_Tumors\normal\normal_128_crop2.png
Saved: Cropped_Tumors\normal\normal_128_crop3.png
Saved: Cropped_Tumors\normal\normal_130_crop0.png
Saved: Cropped_Tumors\normal\normal_130_crop1.png
Saved: Cropped_Tumors\normal\normal_130_crop2.png
Saved: Cropped_Tumors\normal\normal_130_crop3.png
Saved: Cropped_Tumors\normal\normal_132_crop0.png
Saved: Cropped_Tumors\normal\normal_132_crop1.png
Saved: Cropped_Tumors\normal\normal_132_crop2.png
Saved: Cropped_Tumors\normal\normal_132_crop3.png
Saved: Cropped_Tumors\normal\normal_134_crop0.png
Saved: Cropped_Tumors\normal\normal_134_crop1.png
Saved: Cropped_Tumors\normal\normal_134_crop2.png
Saved: Cropped_Tumors\normal\normal_134_crop3.png
Saved: Cropped_Tumors\normal\normal_136_crop0.png
Saved: Cropped_Tumors\normal\normal_136_crop1.png
Saved: Cropped_Tumors\normal\normal_136_crop2.png


 11%|█▏        | 153/1357 [00:02<00:18, 66.74it/s]

Saved: Cropped_Tumors\normal\normal_140_crop0.png
Saved: Cropped_Tumors\normal\normal_140_crop1.png
Saved: Cropped_Tumors\normal\normal_140_crop2.png
Saved: Cropped_Tumors\normal\normal_140_crop3.png
Saved: Cropped_Tumors\normal\normal_142_crop0.png
Saved: Cropped_Tumors\normal\normal_142_crop1.png
Saved: Cropped_Tumors\normal\normal_142_crop2.png
Saved: Cropped_Tumors\normal\normal_142_crop3.png
Saved: Cropped_Tumors\normal\normal_144_crop0.png
Saved: Cropped_Tumors\normal\normal_144_crop1.png
Saved: Cropped_Tumors\normal\normal_144_crop2.png
Saved: Cropped_Tumors\normal\normal_144_crop3.png
Saved: Cropped_Tumors\normal\normal_146_crop0.png
Saved: Cropped_Tumors\normal\normal_146_crop1.png
Saved: Cropped_Tumors\normal\normal_146_crop2.png
Saved: Cropped_Tumors\normal\normal_146_crop3.png
Saved: Cropped_Tumors\normal\normal_148_crop0.png
Saved: Cropped_Tumors\normal\normal_148_crop1.png
Saved: Cropped_Tumors\normal\normal_148_crop2.png
Saved: Cropped_Tumors\normal\normal_148_crop3.png


 12%|█▏        | 163/1357 [00:02<00:16, 72.56it/s]

Saved: Cropped_Tumors\normal\normal_154_crop0.png
Saved: Cropped_Tumors\normal\normal_154_crop1.png
Saved: Cropped_Tumors\normal\normal_154_crop2.png
Saved: Cropped_Tumors\normal\normal_154_crop3.png
Saved: Cropped_Tumors\normal\normal_156_crop0.png
Saved: Cropped_Tumors\normal\normal_156_crop1.png
Saved: Cropped_Tumors\normal\normal_156_crop2.png
Saved: Cropped_Tumors\normal\normal_156_crop3.png
Saved: Cropped_Tumors\normal\normal_158_crop0.png
Saved: Cropped_Tumors\normal\normal_158_crop1.png
Saved: Cropped_Tumors\normal\normal_158_crop2.png
Saved: Cropped_Tumors\normal\normal_158_crop3.png
Saved: Cropped_Tumors\normal\normal_160_crop0.png
Saved: Cropped_Tumors\normal\normal_160_crop1.png
Saved: Cropped_Tumors\normal\normal_160_crop2.png
Saved: Cropped_Tumors\normal\normal_160_crop3.png
Saved: Cropped_Tumors\normal\normal_162_crop0.png
Saved: Cropped_Tumors\normal\normal_162_crop1.png
Saved: Cropped_Tumors\normal\normal_162_crop2.png
Saved: Cropped_Tumors\normal\normal_162_crop3.png


 13%|█▎        | 179/1357 [00:02<00:16, 72.30it/s]

Saved: Cropped_Tumors\normal\normal_170_crop0.png
Saved: Cropped_Tumors\normal\normal_170_crop1.png
Saved: Cropped_Tumors\normal\normal_170_crop2.png
Saved: Cropped_Tumors\normal\normal_170_crop3.png
Saved: Cropped_Tumors\normal\normal_172_crop0.png
Saved: Cropped_Tumors\normal\normal_172_crop1.png
Saved: Cropped_Tumors\normal\normal_172_crop2.png
Saved: Cropped_Tumors\normal\normal_172_crop3.png
Saved: Cropped_Tumors\normal\normal_174_crop0.png
Saved: Cropped_Tumors\normal\normal_174_crop1.png
Saved: Cropped_Tumors\normal\normal_174_crop2.png
Saved: Cropped_Tumors\normal\normal_174_crop3.png
Saved: Cropped_Tumors\normal\normal_176_crop0.png
Saved: Cropped_Tumors\normal\normal_176_crop1.png
Saved: Cropped_Tumors\normal\normal_176_crop2.png
Saved: Cropped_Tumors\normal\normal_176_crop3.png
Saved: Cropped_Tumors\normal\normal_178_crop0.png
Saved: Cropped_Tumors\normal\normal_178_crop1.png
Saved: Cropped_Tumors\normal\normal_178_crop2.png
Saved: Cropped_Tumors\normal\normal_178_crop3.png


 14%|█▍        | 195/1357 [00:02<00:16, 72.31it/s]

Saved: Cropped_Tumors\normal\normal_186_crop2.png
Saved: Cropped_Tumors\normal\normal_186_crop3.png
Saved: Cropped_Tumors\normal\normal_188_crop0.png
Saved: Cropped_Tumors\normal\normal_188_crop1.png
Saved: Cropped_Tumors\normal\normal_188_crop2.png
Saved: Cropped_Tumors\normal\normal_188_crop3.png
Saved: Cropped_Tumors\normal\normal_190_crop0.png
Saved: Cropped_Tumors\normal\normal_190_crop1.png
Saved: Cropped_Tumors\normal\normal_190_crop2.png
Saved: Cropped_Tumors\normal\normal_190_crop3.png
Saved: Cropped_Tumors\normal\normal_192_crop0.png
Saved: Cropped_Tumors\normal\normal_192_crop1.png
Saved: Cropped_Tumors\normal\normal_192_crop2.png
Saved: Cropped_Tumors\normal\normal_192_crop3.png
Saved: Cropped_Tumors\normal\normal_194_crop0.png
Saved: Cropped_Tumors\normal\normal_194_crop1.png
Saved: Cropped_Tumors\normal\normal_194_crop2.png
Saved: Cropped_Tumors\normal\normal_194_crop3.png
Saved: Cropped_Tumors\normal\normal_196_crop0.png
Saved: Cropped_Tumors\normal\normal_196_crop1.png


 16%|█▌        | 211/1357 [00:03<00:15, 74.09it/s]

Saved: Cropped_Tumors\normal\normal_202_crop0.png
Saved: Cropped_Tumors\normal\normal_202_crop1.png
Saved: Cropped_Tumors\normal\normal_202_crop2.png
Saved: Cropped_Tumors\normal\normal_202_crop3.png
Saved: Cropped_Tumors\normal\normal_204_crop0.png
Saved: Cropped_Tumors\normal\normal_204_crop1.png
Saved: Cropped_Tumors\normal\normal_204_crop2.png
Saved: Cropped_Tumors\normal\normal_204_crop3.png
Saved: Cropped_Tumors\normal\normal_206_crop0.png
Saved: Cropped_Tumors\normal\normal_206_crop1.png
Saved: Cropped_Tumors\normal\normal_206_crop2.png
Saved: Cropped_Tumors\normal\normal_206_crop3.png
Saved: Cropped_Tumors\normal\normal_208_crop0.png
Saved: Cropped_Tumors\normal\normal_208_crop1.png
Saved: Cropped_Tumors\normal\normal_208_crop2.png
Saved: Cropped_Tumors\normal\normal_208_crop3.png
Saved: Cropped_Tumors\normal\normal_210_crop0.png
Saved: Cropped_Tumors\normal\normal_210_crop1.png
Saved: Cropped_Tumors\normal\normal_210_crop2.png
Saved: Cropped_Tumors\normal\normal_210_crop3.png


 17%|█▋        | 229/1357 [00:03<00:14, 75.75it/s]

Saved: Cropped_Tumors\normal\normal_220_crop0.png
Saved: Cropped_Tumors\normal\normal_220_crop1.png
Saved: Cropped_Tumors\normal\normal_220_crop2.png
Saved: Cropped_Tumors\normal\normal_220_crop3.png
Saved: Cropped_Tumors\normal\normal_222_crop0.png
Saved: Cropped_Tumors\normal\normal_222_crop1.png
Saved: Cropped_Tumors\normal\normal_222_crop2.png
Saved: Cropped_Tumors\normal\normal_222_crop3.png
Saved: Cropped_Tumors\normal\normal_224_crop0.png
Saved: Cropped_Tumors\normal\normal_224_crop1.png
Saved: Cropped_Tumors\normal\normal_224_crop2.png
Saved: Cropped_Tumors\normal\normal_224_crop3.png
Saved: Cropped_Tumors\normal\normal_226_crop0.png
Saved: Cropped_Tumors\normal\normal_226_crop1.png
Saved: Cropped_Tumors\normal\normal_226_crop2.png
Saved: Cropped_Tumors\normal\normal_226_crop3.png
Saved: Cropped_Tumors\normal\normal_228_crop0.png
Saved: Cropped_Tumors\normal\normal_228_crop1.png
Saved: Cropped_Tumors\normal\normal_228_crop2.png
Saved: Cropped_Tumors\normal\normal_228_crop3.png


 18%|█▊        | 247/1357 [00:03<00:14, 76.57it/s]

Saved: Cropped_Tumors\normal\normal_234_crop0.png
Saved: Cropped_Tumors\normal\normal_234_crop1.png
Saved: Cropped_Tumors\normal\normal_234_crop2.png
Saved: Cropped_Tumors\normal\normal_234_crop3.png
Saved: Cropped_Tumors\normal\normal_236_crop0.png
Saved: Cropped_Tumors\normal\normal_236_crop1.png
Saved: Cropped_Tumors\normal\normal_236_crop2.png
Saved: Cropped_Tumors\normal\normal_236_crop3.png
Saved: Cropped_Tumors\normal\normal_238_crop0.png
Saved: Cropped_Tumors\normal\normal_238_crop1.png
Saved: Cropped_Tumors\normal\normal_238_crop2.png
Saved: Cropped_Tumors\normal\normal_238_crop3.png
Saved: Cropped_Tumors\normal\normal_240_crop0.png
Saved: Cropped_Tumors\normal\normal_240_crop1.png
Saved: Cropped_Tumors\normal\normal_240_crop2.png
Saved: Cropped_Tumors\normal\normal_240_crop3.png
Saved: Cropped_Tumors\normal\normal_242_crop0.png
Saved: Cropped_Tumors\normal\normal_242_crop1.png
Saved: Cropped_Tumors\normal\normal_242_crop2.png
Saved: Cropped_Tumors\normal\normal_242_crop3.png


 19%|█▉        | 263/1357 [00:03<00:15, 68.75it/s]

Saved: Cropped_Tumors\normal\normal_252_crop0.png
Saved: Cropped_Tumors\normal\normal_252_crop1.png
Saved: Cropped_Tumors\normal\normal_252_crop2.png
Saved: Cropped_Tumors\normal\normal_252_crop3.png
Saved: Cropped_Tumors\normal\normal_254_crop0.png
Saved: Cropped_Tumors\normal\normal_254_crop1.png
Saved: Cropped_Tumors\normal\normal_254_crop2.png
Saved: Cropped_Tumors\normal\normal_254_crop3.png
Saved: Cropped_Tumors\normal\normal_256_crop0.png
Saved: Cropped_Tumors\normal\normal_256_crop1.png
Saved: Cropped_Tumors\normal\normal_256_crop2.png
Saved: Cropped_Tumors\normal\normal_256_crop3.png
Saved: Cropped_Tumors\normal\normal_258_crop0.png
Saved: Cropped_Tumors\normal\normal_258_crop1.png
Saved: Cropped_Tumors\normal\normal_258_crop2.png
Saved: Cropped_Tumors\normal\normal_258_crop3.png
Saved: Cropped_Tumors\normal\normal_260_crop0.png
Saved: Cropped_Tumors\normal\normal_260_crop1.png
Saved: Cropped_Tumors\normal\normal_260_crop2.png
Saved: Cropped_Tumors\normal\normal_260_crop3.png


100%|██████████| 1357/1357 [00:03<00:00, 345.04it/s]

Saved: Cropped_Tumors\normal\normal_264_crop0.png
Saved: Cropped_Tumors\normal\normal_264_crop1.png
Saved: Cropped_Tumors\normal\normal_264_crop2.png
Saved: Cropped_Tumors\normal\normal_264_crop3.png
Processing complete. Randomly cropped normal images are saved in: Cropped_Tumors\normal


### Now we retrieve the index of the images of the validation set (that we will use to feed the CNN either through SAM or just by giving the full img)

In [8]:
import re

# Load the CSV file
file_path = "C:/Users/elisa/OneDrive/Desktop/Deep Learning/DL_Project/Deep_Learning_FinalProject/validation_images.csv"  
df_index = pd.read_csv(file_path)

# Function to extract the index from the filename
def extract_index(filename):
    match = re.search(r"tumor_(\d+)_mask[0-2]\.png", filename)  
    if match:
        return int(match.group(1))
    return None  # Return None if no match is found

# Apply the function to extract the index
df_index["extracted_index"] = df_index["filename"].apply(extract_index)

df_index = df_index.drop(columns=['index', 'split'])
df_index

filename category  extracted_index
0     tumor_906_mask1.png   benign            906.0
1     tumor_331_mask0.png   benign            331.0
2     tumor_295_mask0.png   benign            295.0
3     tumor_883_mask0.png   benign            883.0
4     tumor_379_mask0.png   benign            379.0
..                    ...      ...              ...
240   normal_64_crop0.png   normal              NaN
241   normal_50_crop3.png   normal              NaN
242  normal_258_crop1.png   normal              NaN
243  normal_190_crop1.png   normal              NaN
244        normal_174.png   normal              NaN

[245 rows x 3 columns]

In [20]:
#NA are from normal
df_filtered = df_index["extracted_index"]
df_filtered
df_filtered_NA = df_filtered.dropna()
df_filtered_NA

0       906.0
1       331.0
2       295.0
3       883.0
4       379.0
        ...  
107    1281.0
108    1253.0
109    1133.0
110    1059.0
111    1235.0
Name: extracted_index, Length: 112, dtype: float64

In [21]:
df_filtered_rows = df.loc[df_filtered_NA]
df_filtered_rows

image_path  label
906   C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      1
331   C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      1
295   C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      1
883   C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      1
379   C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      1
...                                                 ...    ...
1281  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1253  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1133  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1059  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2
1235  C:/Users/elisa/OneDrive/Desktop/Deep Learning/...      2

[112 rows x 2 columns]

In [22]:
df_filtered_rows.to_csv("full_img_test.csv", index=False)
